### ISSUE TITLE
MQTT failed to send msg. 

### ISSUE THREAD

**Issue by bmrpatel** [Azure/azure-iot-sdks#305]
```
client.sendEvent(msg, function (err) {
  if (err) console.log(err.toString());
});
```

error:
```
TypeError: Cannot read property 'on' of undefined
    at Mqtt.publish (E:\Projects\iot-dev\node_modules\azure-iot-mqtt-base\lib\mqtt.js:130:16)
```

During the debug I see that mqtt.js function `Mqtt.prototype.publish` is called. Then it fails on the line 130 `this.client.on('error', errCallback);` because this.client is undefined. mqtt.js `connect` function should set the value for the `client`, however during debugging I saw that `connect` function is never called. I am just using the sample code for the mqtt protocol. Let me know if you need more info. Thanks

**Comment by pierreca** [comment]
Hi @bmrpatel 
do you call client.open () before calling client.send()? does it succeed?

**Comment by bmrpatel** [comment]
After executing `client.open()`, `client.sendEvent()` is called successfully but there is no message received on iot event hub. I look at DeviceExplorer to monitor all the events received.

I tried that and now the error has been replaced with new issue when I call `client.receive()`. I also tried `client.getReceiver()`. 

```
    client.getReceiver(function (err, rcv) {
           ^

TypeError: client.getReceiver is not a function
    at e:\Projects\iot-dev\device-mqtt.js:23:12
```

Looks like I am doing something wrong and Sample code on the page is not upto date. Can you send me working sample code for mqtt and I will test it and update the sample documented @ [https://github.com/Azure/azure-iot-sdks/tree/master/node/device/transport/mqtt](url)

Here is my code for your reference
```javascript
var connectionString = [my connection string here];

var clientFromConnectionString = require('azure-iot-device-mqtt').clientFromConnectionString;
var Message = require('azure-iot-device').Message;


var client = clientFromConnectionString(connectionString);
client.open(() => {
    console.log('connection opened')

    var msg = new Message('some data from my device');
    client.sendEvent(msg, function (err) {
    if (err) console.log(err.toString());
    });

    //commented to try getReceiver()
    // client.receive(function (err, msg) {
    //   if (err) console.error(err);
    //   else console.log(msg);
    // });

    client.getReceiver(function (err, rcv) {
        rcv.on('message', function (msg) {
            console.log(msg.body);
            rcv.complete(msg, function () {
            console.log('completed');
            });
        });
        rcv.on('errorReceived', function (err) {
            console.warn(err);
        });
    });

});
```

**Comment by pierreca** [comment]
Alright, glad to hear sendEvent is working. Onto the next issue: getReceiver was deprecated in version 1.0.0, in favor of an event pattern on the client.

To receive messages, you now just have to subscribe to the `message` event on the `client` object like so:
```js
client.on('message', function (msg) {
    console.log(msg.body);
    client.complete(msg);
});
```

Caveat: you have to call `open` before you subscribe to the client events for now. Ideally I'd like the user to be able to subscribe to the event as soon as the client is created even if the connection is not opened but that's not entirely working for now (not for all transports).

As an FYI, what went through our head when we decided to go for a breaking change at the time of the public release:
- The user (you) now just have to manage the lifecycle of a single object (`client`) instead of two (`client` + `receiver`)
- The tight coupling between client and receiver made it a pain to handle reconnections (whether automatic or user-driven, if the client is disconnected, the receiver is disconnected too, but we need separate events to tell the user, who will have to reconnect both anyways... long story short having 2 separate objects didn't make sense)
- That's how the other SDKs work.

**Comment by bmrpatel** [comment]
Sounds good. I agree with the change and like the new pattern better. :+1: 

I tried your code and it is receiving message now. However there are 2 issues.

1. `msg.body` is `undefined`. I found that the `msg.data` contains the byte array and it has the message.
2. It is emitting the 'message' event 3 times for each message received. I do call complete and solved the issue with done being undefined, by passing in the callback function to `complete` call.

here is the code:
```
client.on('message', function(msg) {
    console.log(msg.data.toString());
    client.complete(msg, (err) => {
        console.log('completed');
    });
});
```

**Comment by pierreca** [comment]
yes, sorry about that, `msg.data` is the right field to look at.

2 is weird though. I wonder if it could be because the message event is subscribed multiple times? did you experience disconnections/reconnections that would cause client.on('message', ...) to be called multiple times?

**Comment by bmrpatel** [comment]
I tried it couple of different times and it is exactly calling it 3 times and on the device explorer side I only see one success feedback message. Same thing works fine for AMQP protocol. Do you know when is next preview version being released?

**Comment by pierreca** [comment]
new version is out... now :) 1.0.1, those are not preview versions anymore.
I just ran the sample a couple of times over MQTT and got all messages I sent from DeviceExplorer to the device only once. However, I now that version 1.0.0 had a bug on the AMQP transport (now solved) that prevented message completion to happen properly. so the messages you have seen might be from a previous AMQP session too.
Let me know if you still see this "triple message" issue when updating, and if so we'll dig more in the debug traces. if you don't and all is solved, feel free to close the issue or let me know and I'll close it. thanks!

**Comment by bmrpatel** [comment]
I updated to 1.0.1 and still same behavior. Getting the message 4 times as shown in the screenshot bellow.

![image](https://cloud.githubusercontent.com/assets/5115789/13156616/5685fd08-d651-11e5-95f0-61bf9b882df8.png)

**Comment by pierreca** [comment]
can you share your code? I'd like to get a repro on my end

**Comment by bmrpatel** [comment]
Looks like this is working fine now. It must be some other test error. I am getting memory leak error. I will open new issue for that.

Thanks
